fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+    };
+
+    while ps.ptr < ps.length {
+        if ps.is_current_byte(b'{') {
+            if text_element_role == TextElementPosition::LineStart {
+                common_indent = Some(0);
+            }
+            let exp = get_placeable(ps)?;
+            last_non_blank = Some(elements.len());
+            elements.push(PatternElementPointers::Placeable(exp));
+            text_element_role = TextElementPosition::Continuation;
+        } else {
+            let slice_start = ps.ptr;
+            let mut indent = 0;
+            if text_element_role == TextElementPosition::LineStart {
+                indent = ps.skip_blank_inline();
+                if ps.ptr >= ps.length {
+                    break;
+                }
+                let b = ps.source[ps.ptr];
+                if indent == 0 {
+                    if b != b'\n' {
+                        break;
+                    }
+                } else if !ps.is_byte_pattern_continuation(b) {
+                    ps.ptr = slice_start;
+                    break;
+                }
+            }
+            let (start, end, text_element_type, termination_reason) = get_text_slice(ps)?;
+            if start != end {
+                if text_element_role == TextElementPosition::LineStart
+                    && text_element_type == TextElementType::NonBlank
+                {
+                    if let Some(common) = common_indent {
+                        if indent < common {
+                            common_indent = Some(indent);
+                        }
+                    } else {
+                        common_indent = Some(indent);
+                    }
+                }
+                if text_element_role != TextElementPosition::LineStart
+                    || text_element_type == TextElementType::NonBlank
+                    || termination_reason == TextElementTermination::LineFeed
+                {
+                    if text_element_type == TextElementType::NonBlank {
+                        last_non_blank = Some(elements.len());
+                    }
+                    elements.push(PatternElementPointers::TextElement(
+                        slice_start,
+                        end,
+                        indent,
+                        text_element_role,
+                    ));
+                }
+            }
+
+            text_element_role = match termination_reason {
+                TextElementTermination::LineFeed => TextElementPosition::LineStart,
+                TextElementTermination::CarriageReturn => TextElementPosition::Continuation,
+                TextElementTermination::PlaceableStart => TextElementPosition::Continuation,
+                TextElementTermination::EOF => TextElementPosition::Continuation,
+            };
+        }
+    }
+
+    if let Some(last_non_blank) = last_non_blank {
+        let elements = elements
+            .into_iter()
+            .take(last_non_blank + 1)
```

Comment:
Why `+1`?

Reply:
Because of how `take` works. If `last_non_blank` is `0`, we need to take `1`.

Reply:
It keeps the track of which of the `elements` is the last one to be non_blank. Let's imagine this scenario:

```
key = This is
  multiline value

  and it still works


key2 = Value 2
```

For `key` it produces the following TextElements:

```
("This is\n", "   multiline value\n", "\n", "  and it still works\n", "\n", "\n")
```

So, we want to only take the lines `0..3`. It may be that since we filter out blank lines, it would be better to remove that variable and just filter out blank lines. I'll look into this after landing this :)

Reply:
https://github.com/projectfluent/fluent-rs/issues/81